src/Symfony/Component/DependencyInjection/Compiler/AttributeAutoconfigurationPass.php

```diff
@@ -0,0 +1,69 @@
+<?php
+
+/*
+ * This file is part of the Symfony package.
+ *
+ * (c) Fabien Potencier <fabien@symfony.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Symfony\Component\DependencyInjection\Compiler;
+
+use Symfony\Component\DependencyInjection\ChildDefinition;
+use Symfony\Component\DependencyInjection\ContainerBuilder;
+use Symfony\Component\DependencyInjection\Definition;
+
+/**
+ * @author Alexander M. Turek <me@derrabus.de>
+ */
+final class AttributeAutoconfigurationPass implements CompilerPassInterface
+{
+    private $ignoreAttributesTag;
+
+    public function __construct(string $ignoreAttributesTag = 'container.ignore_attributes')
+    {
+        $this->ignoreAttributesTag = $ignoreAttributesTag;
+    }
+
+    public function process(ContainerBuilder $container): void
+    {
+        if (80000 > \PHP_VERSION_ID) {
+            return;
+        }
+
+        $autoconfiguredAttributes = $container->getAutoconfiguredAttributes();
+
+        $definitions = array_filter($container->getDefinitions(), function (Definition $definition) {
+            return $definition->isAutoconfigured()
+                && !$definition->isAbstract()
+                && !$definition->hasTag($this->ignoreAttributesTag)
+                && $definition->getClass();
+        });
+
+        foreach ($definitions as $id => $definition) {
+            if (!$reflector = $container->getReflectionClass($definition->getClass(), false)) {
+                continue;
+            }
+
+            foreach ($reflector->getAttributes() as $attribute) {
+                if (!($configurator = $autoconfiguredAttributes[$attribute->getName()] ?? null)) {
+                    continue;
+                }
+
+                $configurator($childDefinition = new ChildDefinition($id), $attribute->newInstance(), $reflector);
+                $this->merge($definition, $childDefinition);
```

Comment:
The merging logic is already implemented, this should replaced by a call to `$definition->setInstanceOfConditionals()`and everything will happen as expected (eg the current logic is broken with regards to the precedence of explicit configuration.

Reply:
I'll try that.

> Sets the definition templates to conditionally apply on the current definition, keyed by parent interface/class.

I'm wondering what I should pick as keys though, given that an attribute is not really a parent class and I can have multiple templates for a given attribute.

Reply:
This should do it well:
```php
$instanceof = $definition->getInstanceofConditionals();
$conditionals = $instanceof[$reflector->class] ?? new ChildDefinition('');
$configurator($conditionals, $attribute->newInstance(), $reflector);
$instanceof[$reflector->class] = $conditionals;
$definition->setInstanceofConditionals($instanceof);
```

It reuses the same `ChildDefinition` when several attributes are defined, but this is something we must handle anyway: when there are multiple overlapping attributes, we must implement some conflict resolution logic. Reusing the same child-def instance is the simplest conflict-resolution logic we can implement - thus likely the desired one to me.

Reply:
That seems to work, thanks!